cpp/src/barretenberg/honk/transcript/transcript.hpp

```diff
@@ -0,0 +1,267 @@
+    struct RoundData {
+        std::vector<std::string> challenge_label;
+        std::vector<std::pair<std::string, size_t>> entries;
+
+        bool operator==(const RoundData& other) const = default;
+    };
+
+    std::map<size_t, RoundData> manifest;
+
+  public:
+    void print()
+    {
+        for (auto& round : manifest) {
+            info("Round: ", round.first);
+            for (auto& label : round.second.challenge_label) {
+                info("\tchallenge: ", label);
+            }
+            for (auto& entry : round.second.entries) {
+                info("\telement (", entry.second, "): ", entry.first);
+            }
+        }
+    }
+
+    template <typename... Strings> void add_challenge(size_t round, Strings&... labels)
+    {
+        manifest[round].challenge_label = { labels... };
+    }
+    void add_entry(size_t round, std::string element_label, size_t element_size)
+    {
+        manifest[round].entries.emplace_back(element_label, element_size);
+    }
+
+    bool operator==(const TranscriptManifest& other) const = default;
+};
+
+/**
+ * @brief Common transcript functionality for both parties. Stores the data for the current round, as well as the
+ * manifest.
+ *
+ * @tparam FF Field from which we sample challenges.
+ */
+template <typename FF> class BaseTranscript {
+    // TODO(Adrian): Make these tweakable
+    static constexpr size_t HASH_OUTPUT_SIZE = 32;
+    static constexpr size_t MIN_BYTES_PER_CHALLENGE = 128 / 8; // 128 bit challenges
+
+    size_t round_number = 0;
+    std::array<uint8_t, HASH_OUTPUT_SIZE> previous_challenge_buffer{};
+    std::vector<uint8_t> current_round_data;
+
+    // "Manifest" object that records a summary of the transcript interactions
+    TranscriptManifest manifest;
+
+    /**
+     * @brief Compute c_next = H( Compress(c_prev || round_buffer) )
+     *
+     * @return std::array<uint8_t, HASH_OUTPUT_SIZE>
+     */
+    [[nodiscard]] std::array<uint8_t, HASH_OUTPUT_SIZE> get_next_challenge_buffer() const
+    {
+        // Prevent challenge generation if nothing was sent by the prover.
+        ASSERT(!current_round_data.empty());
+
+        // concatenate the hash of the previous round (if not the first round) with the current round data.
+        // TODO(Adrian): Do we want to use a domain separator as the initial challenge buffer?
+        // We could be cheeky and use the hash of the manifest as domain separator, which would prevent us from having
+        // to domain separate all the data. (See https://safe-hash.dev)
+        std::vector<uint8_t> full_buffer;
+        if (round_number > 0) {
+            full_buffer.insert(full_buffer.end(), previous_challenge_buffer.begin(), previous_challenge_buffer.end());
+        }
+        full_buffer.insert(full_buffer.end(), current_round_data.begin(), current_round_data.end());
+
+        // Optionally pre-hash the full buffer to minimize the amount of data passed to the cryptographic hash function.
```

Comment:
It seems like we could alternatively just pass the full buffer to blake3 but we use the (cheaper?) pedersen to compress the potentially large buffer to the size of the pedersen output before doing blake3. I think this is what he meant by "optional". What do you think?

Reply:
Compression is useful when the verifier can efficiently compute group operations. This is not the case with a smart contract verifier where it's cheaper to pass the whole buffer to Keccak. 
It may be optional later on, either with cycle-of-curve recursion, or due to the new hash function we end up using.

Reply:
Thanks for explaining. I updated the comment to this effect.